client/dashboard/utils/domain.ts

```diff
@@ -0,0 +1,60 @@
+import { isAfter, subMinutes } from 'date-fns';
+import { DotcomFeatures } from '../data/constants';
+import { DomainTypes } from '../data/domains';
+import { hasPlanFeature } from './site-features';
+import { userHasFlag } from './user';
+import type { Domain, Site, User } from '../data/types';
+
+export function isRecentlyRegistered( registrationDate: string, numberOfMinutes = 30 ) {
+	return (
+		!! registrationDate &&
+		isAfter( new Date( registrationDate ), subMinutes( new Date(), numberOfMinutes ) )
+	);
+}
+
+export function isDomainRenewable( domain: Domain ) {
+	// Only registered domains can be manually renewed
+	if ( domain.type !== DomainTypes.REGISTERED ) {
+		return false;
+	}
+
+	return (
+		domain.subscription_id &&
+		! domain.pending_renewal &&
+		! domain.pending_registration_at_registry &&
+		! domain.pending_registration &&
+		domain.current_user_can_manage &&
+		( domain.is_renewable || domain.is_redeemable ) &&
+		! domain.aftermarket_auction
+	);
+}
+
+export const shouldUpgradeToMakeDomainPrimary = ( {
+	domain,
+	site,
+	user,
+}: {
+	domain: Domain;
+	site: Site;
+	user: User;
+} ) => {
+	return (
+		( domain.type === DomainTypes.REGISTERED || domain.type === DomainTypes.MAPPED ) &&
+		! domain.current_user_can_create_site_from_domain_only &&
+		! domain.primary_domain &&
+		! domain.wpcom_domain &&
+		! domain.is_wpcom_staging_domain &&
+		userHasFlag( user, 'calypso_allow_nonprimary_domains_without_plan' ) &&
+		!! site.plan?.is_free &&
+		! hasPlanFeature( site, DotcomFeatures.SET_PRIMARY_CUSTOM_DOMAIN )
+	);
+};
+
+export function canSetAsPrimary( domain: Domain, shouldUpgradeToMakePrimary: boolean ): boolean {
```

Comment:
Nit: I know this is existing code, but if it needs `shouldUpgradeToMakePrimary` in all cases, I believe it's this method's responsibility to call it 😄 so it should just accept `domain`, `site`, and `user` params, then call that function internally. The callers will be much cleaner...